google-cloud-pubsub/lib/google/cloud/pubsub/publisher.rb

```diff
@@ -0,0 +1,205 @@
+# Copyright 2017 Google Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+require "monitor"
+require "concurrent"
+require "google/cloud/pubsub/publish_result"
+require "google/cloud/pubsub/service"
+
+module Google
+  module Cloud
+    module Pubsub
+      ##
+      # @private
+      # # Publisher
+      #
+      class Publisher
+        include MonitorMixin
+
+        MAX_BYTES = 5*1024*1024
+        MAX_MESSAGES = 1000
+        TIMER_DELAY = 0.25
```

Comment:
The batching logic is conflating two different durations here.

- The maximum duration the user wants us to buffer a batch while waiting for more messages. This may be arbitrarily long, and is set by the user.
- How frequently we check to see if any duration-based limits have elapsed; this is an implementation detail ("Often enough that no one notices" -- instead of polling with a fixed cadence, you might consider adjusting the execution interval to the next potential expiration). 

By extension: We should not be publishing all batches at the same time. 

More precisely, a batch of messages will be published to a given topic when any of the following conditions is satisfied:

- Number of messages exceeds max_messages
- Number of bytes exceeds max_bytes
- The time elapsed since the first publish call in this batch exceeds max_delay
- User calls publisher.stop/flush/shutdown